torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,146 @@
+            overlap_info.bucket_index_to_bucket[bucket_index] = bucket
+            overlap_info.bucket_index_to_future[bucket_index] = fut
+
+        num_buckets = len(overlap_info.params_per_bucket)
+        # NOTE: The implementation from this point forward assumes that the
+        # buckets are indexed incrementally starting from 0 in the order of
+        # their autograd hooks firing
+        is_last_bucket = bucket_index == num_buckets - 1
+        if not is_last_bucket:
+            return fut
+
+        # Perform partial optimizer step on all buckets after the final
+        # bucket has been computed
+        for bucket_index in range(num_buckets):
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+            num_local_optim_params = len(zero.optim.param_groups[0]["params"])
+            if rank_to_update == rank:
+                gradients: List[Optional[torch.Tensor]] = \
+                    [_NO_PARAM_UPDATE for _ in range(num_local_optim_params)]
+                assert bucket_index in overlap_info.offsets, \
+                    f"Bucket index {bucket_index} was not assigned to rank {rank}"
+                offset = overlap_info.offsets[bucket_index]
+                # Ensure that the all-reduce completes before performing the
+                # the parameter update
+                assert bucket_index in overlap_info.bucket_index_to_future, \
+                    f"All-reduce future for bucket {bucket_index} not saved " \
+                    f"on rank {rank}"
+                allreduce_future = overlap_info.bucket_index_to_future[bucket_index]
+                allreduce_future.wait()
+                bucket_gradients = overlap_info.bucket_index_to_bucket[bucket_index].get_per_parameter_tensors()
+                for i, grad in enumerate(bucket_gradients):
+                    gradients[offset + i] = grad
+                zero._local_step(gradients)
+            device = overlap_info.params_per_bucket[bucket_index][0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            overlap_info.broadcast_handles.append(
+                dist.broadcast(
+                    zero._buckets[device_index][rank_to_update][bucket_index],
+                    src=rank_to_update,
+                    async_op=True
+                )
+            )
+
+        # Zero each parameter's gradient if needed
+        if zero._to_zero_grad:
+            ZeroRedundancyOptimizer._zero_grad(zero._all_params)
+
+        # Ensure that all parameter updates are finished before the
+        # next forward pass
+        _ = list(map(lambda x: x.wait(), overlap_info.broadcast_handles))
+        overlap_info.broadcast_handles.clear()
+
+        # Reset per-iteration information
+        overlap_info.bucket_index_to_future.clear()
+        overlap_info.bucket_index_to_bucket.clear()
+
+        return fut
```

Comment:
This is still the allreduce future, right? Meaning that waiting on this future (by DDP backward()) won't guarantee that the ZeRO broadcast is done? We probably need to return `torch.futures.collect_all(all_zero_broadcast_futures)` for the last bucket instead.

Reply:
Among these reviews, this is the most important one, as it might have impact on correctness.

Reply:
Resolved offline.

Reply:
@andwgu Do you mind describing how this was resolved?

Reply:
https://github.com/andwgu/pytorch/blob/ab54c52fa54dac017ff6c9eadae49a97aa117432/torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py#L175
That line should provide the necessary synchronization already. (The `wait()` on the returned future in `finalize_backward()` will not guarantee that the ZeRO broadcasts are done, but these additional `wait()`s will.)

Also, as a side note, the `torch.futures.collect_all(all_zero_broadcast_futures)` would change the hook semantics since it would then return a future for the broadcasts instead of a future for the bucket all-reduce.